### DISCUSSION TITLE
How to use ClassifierOutputTarget() for a model based on binary classification

### DISCUSSION THREAD

**Discussion by ahxmeds** [jacobgil/pytorch-grad-cam#297] [Q&A] [answered]
I have a model for binary classification that I trained on my own images. The model has a resnet18 backbone with some more dense layers after that and finally ends in a sigmoid function. As a result, the output of my model `out` is a number between (0,1) and I assign an input belonging to the `class = 1` if `out > 0.5`, else if belongs to `class=0`. I am interested in using GradCAM to explain the prediction of the model whenever it predicts `class=1`. When I try to use GradCAM, specifically the line:

   ` grayscale_cam = cam(input_tensor=input,targets=[ClassifierOutputTarget(1)])`

I get an error `IndexError: index 1 is out of bounds for dimension 0 with size 1`. The above line works without error if I use `ClassifierOutputTarget(0)`. I believe this is because I only have one number as my output (at this point, I expect, it should have been a two-numbered output, with the first number giving the probability of input belonging to `class=0` and the second number belonging to `class=1`. 

My question is: what should I assign the `targets` variable while calling the GradCAM object so I am able to explain the prediction of 'class=1'?

My model definition is given below:

```
from torchvision import transforms, models
def get_model():
    # load model
    model = models.resnet18(pretrained=True) 
    for param in model.parameters():
        param.requires_grad = True
    model.avgpool = nn.AdaptiveAvgPool2d(output_size=(1,1))
    model.fc = nn.Sequential(
        nn.Flatten(),
        nn.Linear(512,128),
        nn.ReLU(),
        nn.Dropout(0.2),
        nn.Linear(128,1),
        nn.Sigmoid()
        )
    
    return model 
```

**Comment by jacobgil** [comment, marked as answer]
Hi,
The issue is that we need a target that supports the binary output case.
When we have a binary output, it's already the score of category 1.
So we can create a custom target that multiplies by 1 if we want the second category, and by -1 if we want to first category.
I added a custom target here:
https://github.com/jacobgil/pytorch-grad-cam/blob/master/pytorch_grad_cam/utils/model_targets.py#L21

So you can now just use BinaryClassifierOutputTarget.
